src/metabase/models/card.clj

```diff
@@ -148,15 +179,17 @@
 ;;; --------------------------------------------------- Revisions ----------------------------------------------------
 
 (def ^:private excluded-columns-for-card-revision
-  [:id :created_at :updated_at :entity_id :creator_id :public_uuid :made_public_by_id :metabase_version])
+  [:id :created_at :updated_at :entity_id :creator_id :public_uuid :made_public_by_id :metabase_version
+   ;; we'll use type now
+   :dataset])
 
 (defmethod revision/serialize-instance :model/Card
   ([instance]
    (revision/serialize-instance Card nil instance))
   ([_model _id instance]
    (cond-> (apply dissoc instance excluded-columns-for-card-revision)
      ;; datasets should preserve edits to metadata
-     (not (:dataset instance))
+     (and (:type instance) (not (is-model? instance)))
```

Comment:
Why not keep backwards compatibility here, like other places in the code? `(or (is-model? instance) (:dataset instance))`. This version feels strange - checking that :type is present (i.e. not assuming the migration) but then not handling :dataset seems unbalanced.

Reply:
dataset is no longer serialized for revision, you can see in line 184 I added `:dataset` to the list of `excluded-columns-for-card-revision`

Reply:
Why is this not `(not (is-model? instance))`?

Reply:
Bumping this. Including the `(:type instance)` is unnecessary imo. It implies it's possible to have `instance` without a `:type` key, and implies that case should be handled differently than if `(:type instance)` was present but `(not (is-model? instance))` is false. But it shouldn't be possible to have `instance` without a `:type` key here because `revision/serialize-instance` should be called with `instance` being a selected toucan object from the database after insertion/update.

Reply:
ah yea, the `(:type instance)` is only needed in revision tests because there are tests where the object will only contain a key that is needed to generate a revision description.

for example: https://github.com/metabase/metabase/blob/metric-v2/add-report-card-type-shortcut/test/metabase/models/revision_test.clj#L341

I'll added a comment to the code.